### PR TITLE
Created test file for search.js with example tests.

### PR DESCRIPTION
**Pull request by carmenBen**
Added mocha, chai, and sinon testing libraries. Created sample tests.

Doc with information on testing.
https://docs.google.com/document/d/1ixjlfk8JPncxnM19FOEmD_WnU65JqIT-TQuVO3P647w/edit?usp=sharing

### CODE REVIEW

File: search-sentiments-map/routes/search.js [hunk 1]
```diff
@@ -272,11 +273,33 @@ async function addWorldDataByTopicToDatastore(topic, countriesData) {
...
 
-module.exports.router = router;
-module.exports.updateSearchResults = updateSearchResults;
\ No newline at end of file
+/** Sleep for one minute. */
+function sleepForOneMinute() {
+  return new Promise(resolve => setTimeout(resolve, PAUSE_ONE_MIN_MS));
+}
+
+// Necessary for t
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/routes/search.js:289]: Unfinished comment?

File: search-sentiments-map/routes/search.js [hunk 2]
```diff
@@ -200,7 +201,7 @@ function formatSearchResult(searchResult) {
  * @param {Object} searchResult Object for one search result.
  */
 function getSentiment(searchResult) {
-  return fetch('https://trending-search-sentiments.ue.r.appspot.com/sentiment', {
+  return fetch('https://trending-search-sentiments.ue.r.appspot.com/sentiment',{
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/routes/search.js:204]: why are you removing the space?

File: search-sentiments-map/routes/search.js [hunk 3]
```diff
@@ -141,12 +141,13 @@ async function updateSearchResultsForTopic(query) {
...
   const {searchApiKey} = require('./config.js');
   let response = 
-      await fetch('https://www.googleapis.com/customsearch/v1?key=' + searchApiKey
-          + '&cx=017187910465527070415:o5pur9drtw0&q='  + query
+      await fetch('https://www.googleapis.com/customsearch/v1?key='
+          + searchApiKey + '&cx=017187910465527070415:o5pur9drtw0&q='  + query
           + '&cr=country' + countryCode
-          + '&num=10&safe=active&dateRestrict=d1&fields=items(title,snippet,htmlTitle,link)');
+          + '&num=10&safe=active&dateRestrict=d1&fields=items'
+          + '(title,snippet,htmlTitle,link)');
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/routes/search.js:148]: Maybe line wrap at the &?

File: search-sentiments-map/test/search-test.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+          sinon.stub(search, 'deleteAncientResults')
+              .resolves('Not interested in the output');
+      sleepForOneMinuteStub = 
+          sinon.stub(search, 'sleepForOneMinute')
+              .resolves('Not interested in the output');
+      retrieveGlobalTrendsStub = 
+          sinon.stub(search, 'retrieveGlobalTrends')
+              .resolves([
+                {trendTopic: "trend1"},
+                {trendTopic: "trend2"}
```
Comment (ChenyuHeidiZhang) [inline search-sentiments-map/test/search-test.js:25]: single quotes

Comment (ChenyuHeidiZhang) [reply search-sentiments-map/test/search-test.js:25]: comma at the end

File: search-sentiments-map/routes/search.js (hunk 1 above)
Comment (carmenBen) [reply search-sentiments-map/routes/search.js]: Sometime the file hasn't fully saved by the time I push(quite a lag at times). Fixed it.

File: search-sentiments-map/routes/search.js [hunk 5]
```diff
@@ -71,20 +71,20 @@ async function retrieveSearchResultFromDatastore(topic) {
...
-  });
+    await search.sleepForOneMinute();
+  }
 }
 
 /** 
  * Obtains the most recent global trends by querying the Datastore.
- * @return {!Array<JSON>} A JSON array of global trends and their originating countries.
+ * @return {!Array<JSON>} A JSON array of global trends and their originating
+ *     countries.
```
Comment (ntarn) [inline search-sentiments-map/routes/search.js:87]:
```suggestion
 * countries.
```
same styling issue as in the previous PR for block comments, JSDOc comments

File: search-sentiments-map/routes/search.js [hunk 6]
```diff
@@ -141,12 +141,13 @@ async function updateSearchResultsForTopic(query) {
 async function getCustomSearchResultsForCountry(countryCode, query) {
   const {searchApiKey} = require('./config.js');
   let response = 
-      await fetch('https://www.googleapis.com/customsearch/v1?key=' + searchApiKey
-          + '&cx=017187910465527070415:o5pur9drtw0&q='  + query
+      await fetch('https://www.googleapis.com/customsearch/v1?key='
+          + searchApiKey + '&cx=017187910465527070415:o5pur9drtw0&q='  + query
```
Comment (ntarn) [inline search-sentiments-map/routes/search.js:145]:
```suggestion
          + searchApiKey + '&cx=017187910465527070415:o5pur9drtw0&q=' + query
```

File: search-sentiments-map/routes/search.js (hunk 5 above)
Comment (carmenBen) [reply search-sentiments-map/routes/search.js:87]: See comment from Chenyu on the other comment. Section 7.4 of the Javascript style guide shows this is correct.

File: search-sentiments-map/routes/search.js [hunk 7]
```diff
@@ -272,11 +274,33 @@ async function addWorldDataByTopicToDatastore(topic, countriesData) {
...
   } catch (err) {
     console.error('ERROR:', err);
   }
 }
 
-module.exports.router = router;
-module.exports.updateSearchResults = updateSearchResults;
\ No newline at end of file
+/** Sleep for one minute. */
+function sleepForOneMinute() {
```
Comment (hayleejane3) [inline search-sentiments-map/routes/search.js:286]: just `sleepOneMin` should be good enough.

Comment (hayleejane3) [inline search-sentiments-map/routes/search.js:286]: Also how about a generic sleep where you can pass in a param according to the time you need

File: search-sentiments-map/routes/search.js [hunk 8]
```diff
@@ -272,11 +274,33 @@ async function addWorldDataByTopicToDatastore(topic, countriesData) {
...
+  updateSearchResults,
+  retrieveGlobalTrends,
+  updateSearchResultsForTopic,
+  getCustomSearchResultsForCountry,
+  formatCountryResults,
+  formatSearchResult,
+  getSentiment,
+  deleteAncientResults,
+  addWorldDataByTopicToDatastore,
+  sleepForOneMinute,
```
Comment (hayleejane3) [inline search-sentiments-map/routes/search.js:302]: If the order doesn't matter here I'd recommend grouping them by similar in concept/what they do/handle

File: search-sentiments-map/test/search-test.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+      updateSearchResultsForTopicStub =
+          sinon.stub(search, 'updateSearchResultsForTopic')
+              .resolves('Not interested in the output');
+    })
+
+    afterEach(() => {
+      sinon.restore();
+    })
+
+    it('should update search results for all global trends', async function() {
```
Comment (hayleejane3) [inline search-sentiments-map/test/search-test.js:36]:
```suggestion
    it('Should update search results for all global trends.', async function() {
```

File: search-sentiments-map/test/search-test.js [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+      await search.updateSearchResults();
+      deleteAncientResultsStub.callCount.should.equal(1);
+      retrieveGlobalTrendsStub.callCount.should.equal(1);
+      updateSearchResultsForTopicStub.getCall(0).args[0].should.equal('trend1');
+      updateSearchResultsForTopicStub.getCall(1).args[0].should.equal('trend2');
+      updateSearchResultsForTopicStub.callCount.should.equal(2);
+      sleepForOneMinuteStub.callCount.should.equal(2);
```
Comment (hayleejane3) [inline search-sentiments-map/test/search-test.js:43]: Make sure that these checks are in the order that the calls would be expected to be made.

Also, you would ideally have these checks immediately after the code that would call each of these.

File: search-sentiments-map/test/search-test.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        }];
+        return [results];
+      });
+    });
+
+    afterEach(() => {
+      sinon.restore();
+    });
+
+    it('should return latest datastore data', async function() {
```
Comment (hayleejane3) [inline search-sentiments-map/test/search-test.js:63]:
```suggestion
    it('Should return latest datastore data.', async function() {
```

Comment (carmenBen) [reply search-sentiments-map/test/search-test.js:63]: It seems that mocha convention is for the lowercase, no full stop formatting. I can change this to your suggestion, but I just wanted to check.

https://mochajs.org/#getting-started

File: search-sentiments-map/test/search-test.js (hunk 9 above)
Comment (carmenBen) [reply search-sentiments-map/test/search-test.js:36]: See comment below.

File: search-sentiments-map/test/search-test.js (hunk 10 above)
Comment (carmenBen) [reply search-sentiments-map/test/search-test.js:43]: I have not seen anyway for this to be called immediately after the code calls each of these. These are in order of their calls.

File: search-sentiments-map/routes/search.js (hunk 8 above)
Comment (carmenBen) [reply search-sentiments-map/routes/search.js:302]: This is currently in the same order these functions are declared in the file. Is that fine? They are all fairly connected to each other so I am not super sure how well I could order them some other way. I could also do alphabetical order.

File: search-sentiments-map/test/search-test.js (hunk 11 above)
Comment (hayleejane3) [reply search-sentiments-map/test/search-test.js:63]: Its fine as is :)

**Pull request merged**